irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -46,85 +46,160 @@ namespace iroha {
+          logger::Logger log) {
+        return mutable_factory->createMutableStorage().match(
+            [](expected::Value<std::unique_ptr<ametsuchi::MutableStorage>>
+                   &created_storage) {
+              return std::move(created_storage.value);
+            },
+            [&log](expected::Error<std::string> &error) {
+              log->error("could not create mutable storage: {}", error.error);
+              return std::unique_ptr<ametsuchi::MutableStorage>{};
+            });
+      }
+
+      /**
+       * Process block, which can be applied to current storage directly:
+       *   - apply non-empty block and commit result to Ametsuchi
+       *     @or
+       *   - don't apply empty block
+       * In both cases notify the subscriber about commit
+       * @param committed_block_variant to be applied
+       * @param notifier to notify the pcs about commit
+       * @param mutable_factory to create storage for block application
+       * @param log to tell about errors
+       */
+      void processApplicableBlock(
+          const shared_model::interface::BlockVariant &committed_block_variant,
+          const rxcpp::subjects::subject<Commit> &notifier,
+          std::shared_ptr<ametsuchi::MutableFactory> mutable_factory,
+          logger::Logger log) {
+        iroha::visit_in_place(
+            committed_block_variant,
+            [&](std::shared_ptr<shared_model::interface::Block> block_ptr) {
+              auto applyStorage = createTemporaryStorage(mutable_factory, log);
+              if (not applyStorage) {
+                return;
+              }
+              applyStorage->apply(*block_ptr, trueStorageApplyPredicate);
+              mutable_factory->commit(std::move(applyStorage));
+
+              notifier.get_subscriber().on_next(
+                  rxcpp::observable<>::just(block_ptr));
+            },
+            [&](std::shared_ptr<shared_model::interface::EmptyBlock>
+                    empty_block_ptr) {
+              notifier.get_subscriber().on_next(
+                  rxcpp::observable<>::empty<
+                      std::shared_ptr<shared_model::interface::Block>>());
+            });
+      }
+
+      /**
+       * Process block, which cannot be applied to current storage directly:
+       *   - try to download missing blocks from other peers (don't stop, if
+       *     they cannot provide blocks at that moment)
+       *   - apply the chain on top of existing storage and commit result
+       * Committed block variant is not applied, because it's either empty or
+       * already exists in downloaded chain
+       * @param committed_block_variant to be processed
+       * @param notifier to notify the pcs about commit
+       * @param mutable_factory to commit results
+       * @param storage to validate and apply chain
+       * @param block_loader to download the blocks
+       * @param validator to validate chain
+       */
+      void processUnapplicableBlock(
```

Comment:
Please split this function, it is hard to understand. I think the algorithm you outlined in the comments can be used to implement a separate function for each state